### ISSUE TITLE
Reflexive constraint involving erased projection not solved

### ISSUE THREAD

**Issue by nad** [agda/agda#4137]
I have some code that used to work. Bisection suggests that, starting from b3aec6fa44a5ef077c30e563208c2274ef39e4d4 ("[ #3855 ] occursCheck: check for erased definitions"), the code fails to type-check. Apparently there is a constraint of the form "e = e" that Agda refuses to solve. Further bisection suggests that, starting from 72c8afd53b7b3e6b0af2b73fb6066f9e4140fdfd ("[ #4125 ] Use unreduced type in checkExpr"), the code (also?) triggers an internal error:
```
An internal error has occurred. Please report this as a bug.
Location of the error: src/full/Agda/TypeChecking/Substitute.hs:83
```

It might be possible to reproduce the behaviour mentioned above in the following way (using a suitable version of Agda):
```
git clone http://www.cse.chalmers.se/~nad/repos/equality/
cd equality
git checkout 555979e856d1d7a12c15ce4bd342d5f19522e4b9
agda src/Queue/Truncated/Instances.agda
```

**Comment by nad** [comment]
The following code triggers the internal error:
```agda
open import Agda.Primitive

postulate

  A : Set
  F : ∀ {ℓ} → Set ℓ → Set ℓ

record R (Q : ∀ {ℓ} → Set ℓ → Set ℓ) ℓ : Set (lsuc ℓ) where
  field
    f : {A : Set ℓ} → F A → Q A
    g : {A : Set ℓ} → Q A → F A

open R ⦃ … ⦄

module M (_ : Set₁) where

  postulate

    G : ∀ {ℓ} → Set ℓ → Set ℓ

module _ (_ : Set) where

  open M Set

  postulate

    instance

      _ : R G lzero

  g′ : G A → F A
  g′ = g

  variable

    x : F A

  postulate

    P     : F A → Set
    works : {x : F A} → P (g′ (f x))
    fails : P (g′ (f x))
```
This code is accepted by b3aec6fa44a5ef077c30e563208c2274ef39e4d4, so perhaps the two issues are separate.

**Comment by nad** [comment]
Here is a test case for the other issue:
```agda
{-# OPTIONS --show-implicit #-}

open import Agda.Builtin.List

foldr : {A B : Set} → (A → B → B) → B → List A → B
foldr _⊕_ ε []       = ε
foldr _⊕_ ε (x ∷ xs) = x ⊕ foldr _⊕_ ε xs

infixr 5 _++_

_++_ : {A : Set} → List A → List A → List A
xs ++ ys = foldr _∷_ ys xs

record R (F : Set → Set) : Set₁ where
  field
    f : {A : Set} → A → F A → F A

open R ⦃ … ⦄ public

postulate
  D    : {A : Set} → List A → Set
  easy : {A : Set} {@0 xs : List A} → D xs

record Q (A : Set) : Set where
  field
    @0 index : List A
    d        : D index

g : {A : Set} → A → Q A → Q A
g x q .Q.index = q .Q.index ++ x ∷ []
g x q .Q.d     = easy

record _×_ (A B : Set) : Set where
  constructor _,_
  field
    proj₁ : A
    proj₂ : B

data Maybe (A : Set) : Set where
  nothing : Maybe A
  just    : A → Maybe A

postulate
  m₁ : {A : Set} {B : Maybe A → Set} →
       ((x : A) → B (just x)) → (x : Maybe A) → B x

m₂ : {A B : Set} → (A → Maybe B) → Maybe A → Maybe B
m₂ f x = m₁ f x

postulate
  P : {A : Set} → A → Set
  p : {A : Set} (x : A) → P x
  A : Set
  x : Maybe (A × Q A)

instance
  _ : R Q
  _ = record { f = g }

_ : P (m₂ (λ { (x , q) → just (f x q) }) x)
_ = p (m₁ (λ { (x , q) → just (f x q) }) x)
```
Agda fails to solve a constraint of the form `e = e`:
```
Failed to solve the following constraints:
  q .Q.index = q .Q.index : List {Agda.Primitive.lzero} A
```

**Comment by andreasabel** [comment]
The first problem does not involve erasure, so I split it out to #4149.  
The second problem likely is caused by erasure, as the problematic terms involve the erased projection `index`.

**Comment by UlfNorell** [comment]
It isn't actually a reflexive constraint. One side is a blocked term, but we unblock those when printing constraints. The reason it doesn't get instantiated is the `definitionCheck` in `MetaVars.Occurs`. Possibly the irrelevance/erasure of the blocked term meta is not set correctly?
